### DISCUSSION TITLE
Automatic renewal for self-issued certificate

### DISCUSSION THREAD

**Discussion by georgiosd** [ravendb/ravendb#14604] [Q&A]
I have a ticket open with support, but I thought more people would benefit from this discussion.

So, we are using a custom domain for our cluster and while setting it up, we issued a certificate using let's encrypt. We now want Raven to automatically update it when it's time.

Reading https://ravendb.net/docs/article-page/5.3/csharp/server/security/authentication/lets-encrypt-certificates#automatic-renewal-for-self-obtained-certificates, I expected that if I add:

```
 "Security.Certificate.Load.Exec": "/bin/bash",
 "Security.Certificate.Load.Exec.Arguments": "renew-cert.sh"
```
and then I go  to Manage Server => Certificates, I will see a "Renew" button.

I now understand the button is only there for certificates that Raven has issued - but if Raven knows what to do to renew the cert, shouldn't it always be there?

Button aside, while Raven does seem to visit my script every so often, it doesn't seem to execute it - what I see in the logs is:

2022-07-23T06:26:58.7477418Z, 1, Operations, Server, Raven.Server.RavenServer, You are using the configuration property 'Security.Certificate.Load.Exec', without specifying 'Security.Certificate.Renew.Exec' and 'Security.Certificate.Change.Exec'. This configuration requires you to renew the certificate manually across the entire cluster.

This part followed but not sure if it's relevant:

2022-07-23T06:27:00.3367050Z, 1, Operations, Server, Raven.Server.RavenServer, When setting the certificate, validating that the server can authenticate with itself using https://host/.
2022-07-23T06:27:06.8295081Z, 33, Operations, Server, Raven.Server.RavenServer, Successful connection to https://host/.

I did add logging to the script itself but it's not getting written, so I think it never gets executed.

**Comment by ayende** [comment]
> The certificate path setting takes precedence over executable configuration option.

Are you defining both of them?

Note that there is distinction between:
* `Security.Certificate.Load.Exec` - read the cert from somewhere. Can be certbot, HSM, etc.
* `Security.Certificate.Change.Exec` - *accepts* a certificate and save it.

And:
* `Security.Certificate.Renew.Exec` - will be invoked once an hour by RavenDB to check if you need a new cert
 
The two aren't necessarily connected, and are able to handle pretty complex setup.

In general, if you just want to renew with custom cert, just use the renew option.
RavenDB will store the cert on the file system.

If you want to store the certificates on an Hardware Security Module or similar (like Azure Vault), you can use the Load & Change scripts to read & write it.

Note that a lot depends if all nodes read from the same location or not. But for your scenario, I would say you just want `renew`

**Reply by georgiosd** [reply]
Success with `Renew`, thank you!

Would you agree this scenario deserves a bit more attention in the docs?

**Reply by ayende** [reply]
Yes, I created an issue for that.

**Reply by ayende** [reply]
https://issues.hibernatingrhinos.com/issue/RDoc-2236/Cover-the-renew-custom-certificate-automatically-in-more-depth

**Reply by panagushin** [reply]
Looks like solution is undergoing?

**Reply by georgiosd** [reply]
@OlegPanagushin Here's mine:

renew-cert.sh
```
#!/bin/bash

set -o errexit

PUBLIC_SERVER_URL=`jq -r '.PublicServerUrl' ./settings.json`
NODE=`echo "$PUBLIC_SERVER_URL" | awk -F[/:] '{print $4}'`
DOMAIN="${NODE#*.}"
HOST="*.$DOMAIN"
USER=`whoami`
HOME="/home/$USER"
CERTBOT="$HOME/.certbot"
CERTDIR="$CERTBOT/config/live/$DOMAIN"
LOG="renew-cert.log"

echo "Generating certificate for $HOST" > $LOG

certbot -d $HOST certonly \
    --config-dir $CERTBOT/config \
    --logs-dir $CERTBOT/logs \
    --work-dir $CERTBOT/work \
    --dns-google \
    --dns-google-credentials $HOME/.secrets/certbot-dns.json \
    --non-interactive \
    --agree-tos \
    -m email@zerotomvp.dev >> $LOG 2>&1

echo "Success!" >> $LOG

# https://stackoverflow.com/a/62863490/165656
openssl pkcs12 \
    -export \
    -keypbe NONE \
    -certpbe NONE \
    -nomaciter \
    -passout pass: \
    -out $CERTDIR/cert.pfx \
    -inkey $CERTDIR/privkey.pem \
    -in $CERTDIR/cert.pem \
    -certfile $CERTDIR/fullchain.pem >> $LOG 2>&1

echo "Combined to PFX successfully." >> $LOG

cat -u $CERTDIR/cert.pfx
```
Then in `/path/to/RavenDB/Server/settings.json`:
```
{
  ...
  "Security.Certificate.Renew.Exec": "/bin/bash",
  "Security.Certificate.Renew.Exec.Arguments": "renew-cert.sh"
}
```
Note that this is using the Google DNS plugin for the verification and requires `jq` to be available on the host.

**Comment by haludi** [comment]
Only the documentation.
You can check also -
https://ravendb.net/docs/article-page/5.4/csharp/server/security/authentication/certificate-configuration#with-logic-foreign-to-ravendb-or-external-certificate-storage
